### PR TITLE
Enable CSS linting through stylelint

### PR DESCRIPTION
**Pull request by jelly**

### CODE REVIEW

**Comment by jelly** [comment]
```
ERROR in [stylelint]
pkg/sosreport/sosreport.scss
  7:5   ✖  Expected indentation of 2 spaces  indentation
  8:5   ✖  Expected indentation of 2 spaces  indentation
 12:5   ✖  Expected indentation of 2 spaces  indentation
 15:9   ✖  Unexpected whitespace after "("   media-feature-parentheses-space-inside
 15:51  ✖  Unexpected whitespace before ")"  media-feature-parentheses-space-inside
 16:5   ✖  Expected indentation of 2 spaces  indentation
 17:9   ✖  Expected indentation of 4 spaces  indentation
 18:5   ✖  Expected indentation of 2 spaces  indentation
 21:9   ✖  Unexpected whitespace after "("   media-feature-parentheses-space-inside
 21:47  ✖  Unexpected whitespace before ")"  media-feature-parentheses-space-inside
 22:5   ✖  Expected indentation of 2 spaces  indentation
 23:9   ✖  Expected indentation of 4 spaces  indentation
 24:5   ✖  Expected indentation of 2 spaces  indentation

webpack 5.72.1 compiled with 1 error in 12227 ms
CSSLINT=1 ./tools/webpack-make sosreport  24.73s user 0.75s system 189% cpu 13.424 total
[jelle@t14s][~/projects/cockpit/stylelint]%time CSSLINT=0 ./tools/webpack-make sosreport
CSSLINT=0 ./tools/webpack-make sosreport  22.56s user 0.58s system 183% cpu 12.593 total
```

**Comment by garrett** [comment]
I found all the autofixing rules according to the documentation, made a template file that disabled all the rules (after importing the base ones) and wrote a 1-liner loop for fish shell (the shell I use) that iterated through the listing of the commands, excluded the line from the template file, ran stylelint with the custom config with `--fix`, and did a git commit with a basic summary.

Fish loop:

```fish
for i in (cat autorules); echo $i; grep -v \"$i\" stylelint-template.json > rules.json; npx stylelint --config rules.json --fix "pkg/**/*.*css"; git commit -a -m "stylelint: $i"; end
```

Template to turn off things:

```json
{
  "extends": "stylelint-config-standard-scss",
  "rules": {
	"alpha-value-notation": null,
	"at-rule-empty-line-before": null,
	"at-rule-name-case": null,
	"at-rule-name-space-after": null,
	"at-rule-no-vendor-prefix": null,
	"at-rule-semicolon-newline-after": null,
	"block-closing-brace-empty-line-before": null,
	"block-closing-brace-newline-after": null,
	"block-closing-brace-newline-before": null,
	"block-closing-brace-space-before": null,
	"block-opening-brace-newline-after": null,
	"block-opening-brace-newline-before": null,
	"block-opening-brace-space-after": null,
	"block-opening-brace-space-before": null,
	"color-function-notation": null,
	"color-hex-case": null,
	"color-hex-length": null,
	"comment-empty-line-before": null,
	"comment-whitespace-inside": null,
	"custom-property-empty-line-before": null,
	"declaration-bang-space-after": null,
	"declaration-bang-space-before": null,
	"declaration-block-semicolon-newline-after": null,
	"declaration-block-semicolon-space-after": null,
	"declaration-block-semicolon-space-before": null,
	"declaration-block-trailing-semicolon": null,
	"declaration-colon-newline-after": null,
	"declaration-colon-space-after": null,
	"declaration-colon-space-before": null,
	"declaration-empty-line-before": null,
	"font-family-name-quotes": null,
	"function-comma-newline-after": null,
	"function-comma-newline-before": null,
	"function-comma-space-after": null,
	"function-comma-space-before": null,
	"function-max-empty-lines": null,
	"function-name-case": null,
	"function-parentheses-newline-inside": null,
	"function-parentheses-space-inside": null,
	"function-whitespace-after": null,
	"hue-degree-notation": null,
	"indentation": null,
	"length-zero-no-unit": null,
	"linebreaks": null,
	"max-empty-lines": null,
	"media-feature-colon-space-after": null,
	"media-feature-colon-space-before": null,
	"media-feature-name-case": null,
	"media-feature-name-no-vendor-prefix": null,
	"media-feature-parentheses-space-inside": null,
	"media-feature-range-operator-space-after": null,
	"media-feature-range-operator-space-before": null,
	"media-query-list-comma-newline-after": null,
	"media-query-list-comma-space-after": null,
	"media-query-list-comma-space-before": null,
	"no-empty-first-line": null,
	"no-eol-whitespace": null,
	"no-extra-semicolons": null,
	"no-missing-end-of-source-newline": null,
	"number-leading-zero": null,
	"number-no-trailing-zeros": null,
	"property-case": null,
	"property-no-vendor-prefix": null,
	"rule-empty-line-before": null,
	"selector-attribute-brackets-space-inside": null,
	"selector-attribute-operator-space-after": null,
	"selector-attribute-operator-space-before": null,
	"selector-combinator-space-after": null,
	"selector-combinator-space-before": null,
	"selector-descendant-combinator-no-non-space": null,
	"selector-list-comma-newline-after": null,
	"selector-list-comma-newline-before": null,
	"selector-list-comma-space-after": null,
	"selector-list-comma-space-before": null,
	"selector-max-empty-lines": null,
	"selector-no-vendor-prefix": null,
	"selector-not-notation": null,
	"selector-pseudo-class-case": null,
	"selector-pseudo-class-parentheses-space-inside": null,
	"selector-pseudo-element-case": null,
	"selector-pseudo-element-colon-notation": null,
	"selector-type-case": null,
	"shorthand-property-no-redundant-values": null,
	"string-quotes": null,
	"unit-case": null,
	"value-keyword-case": null,
	"value-list-comma-newline-after": null,
	"value-list-comma-space-after": null,
	"value-list-comma-space-before": null,
	"value-list-max-empty-lines": null,
	"value-no-vendor-prefix": null
  }
}
```

All the autofixing rules:

```
alpha-value-notation
at-rule-empty-line-before
at-rule-name-case
at-rule-name-space-after
at-rule-no-vendor-prefix
at-rule-semicolon-newline-after
block-closing-brace-empty-line-before
block-closing-brace-newline-after
block-closing-brace-newline-before
block-closing-brace-space-before
block-opening-brace-newline-after
block-opening-brace-newline-before
block-opening-brace-space-after
block-opening-brace-space-before
color-function-notation
color-hex-case
color-hex-length
comment-empty-line-before
comment-whitespace-inside
custom-property-empty-line-before
declaration-bang-space-after
declaration-bang-space-before
declaration-block-semicolon-newline-after
declaration-block-semicolon-space-after
declaration-block-semicolon-space-before
declaration-block-trailing-semicolon
declaration-colon-newline-after
declaration-colon-space-after
declaration-colon-space-before
declaration-empty-line-before
font-family-name-quotes
function-comma-newline-after
function-comma-newline-before
function-comma-space-after
function-comma-space-before
function-max-empty-lines
function-name-case
function-parentheses-newline-inside
function-parentheses-space-inside
function-whitespace-after
hue-degree-notation
indentation
length-zero-no-unit
linebreaks
max-empty-lines
media-feature-colon-space-after
media-feature-colon-space-before
media-feature-name-case
media-feature-name-no-vendor-prefix
media-feature-parentheses-space-inside
media-feature-range-operator-space-after
media-feature-range-operator-space-before
media-query-list-comma-newline-after
media-query-list-comma-space-after
media-query-list-comma-space-before
no-empty-first-line
no-eol-whitespace
no-extra-semicolons
no-missing-end-of-source-newline
number-leading-zero
number-no-trailing-zeros
property-case
property-no-vendor-prefix
rule-empty-line-before
selector-attribute-brackets-space-inside
selector-attribute-operator-space-after
selector-attribute-operator-space-before
selector-combinator-space-after
selector-combinator-space-before
selector-descendant-combinator-no-non-space
selector-list-comma-newline-after
selector-list-comma-newline-before
selector-list-comma-space-after
selector-list-comma-space-before
selector-max-empty-lines
selector-no-vendor-prefix
selector-not-notation
selector-pseudo-class-case
selector-pseudo-class-parentheses-space-inside
selector-pseudo-element-case
selector-pseudo-element-colon-notation
selector-type-case
shorthand-property-no-redundant-values
string-quotes
unit-case
value-keyword-case
value-list-comma-newline-after
value-list-comma-space-after
value-list-comma-space-before
value-list-max-empty-lines
value-no-vendor-prefix
```

**Comment by garrett** [comment]
The alpha-value-notation entries are wrong, as the first is from a test run that committed the node modules and the second caught some additional auto fixing rules that weren't disabled (as I guess the documentation didn't cover everything).

**Comment by jelly** [comment]
```
[jelle@t14s][~/projects/cockpit/add_stylelint]%hyperfine --warmup 3 'STYLELINT=0 ./tools/webpack-make systemd' && hyperfine --warmup 3 'STYLELINT=1 ./tools/webpack-make systemd'
Benchmark 1: STYLELINT=0 ./tools/webpack-make systemd
  Time (mean ± σ):     27.338 s ±  0.301 s    [User: 46.508 s, System: 1.162 s]
  Range (min … max):   26.845 s … 27.787 s    10 runs

Benchmark 1: STYLELINT=1 ./tools/webpack-make systemd
  Time (mean ± σ):     27.938 s ±  0.445 s    [User: 47.845 s, System: 1.192 s]
  Range (min … max):   27.326 s … 28.763 s    10 runs
```

**Comment by martinpitt** [comment]
PR #17471 laid the foundation for this (just landed). Now we can resurrect this PR and land this piece by piece.

**Comment by garrett** [comment]
Reopened. It doesn't rebase cleanly, and it does conflict with some login.css changes already, and will conflict with more once the dark login PR lands.

Suggested path forward: I'll wait until the dark login changes land, reset hard to origin/main, then re-run the script as outlined above to reproduce the commits.

**Comment by garrett** [comment]
There _seems_ to be a default ruleset that cannot be turned off when doing the fixing. You can see it at https://github.com/cockpit-project/cockpit/pull/17408/commits/6efb0fccdfb9f89ddf83ced5c26d60007a7b1715.

I re-ran it with everything and rebased my previous fixes on top of the stack. After all the cleanups, I further adjusted the rc file to drop the rules that aren't hit and only keep the ones that are.

The rules that are left are things I should go through manually to fix. If they shouldn't be fixed (we're likely to have a few), then they should stick around as ignores and/or should be handled by in-line comments.

I still need to go through all the commits again for a more throough review. And I should probably squash the manual changes I've made into related auto-commits if/when they exist.

Also, I wonder what's going to happen with our pixel tests.

**Comment by garrett** [comment]
"Default" ruleset adds spaces above comments (which I think isn't really right), strips spaces from the color parts (which makes it easier to read before and stylelint makes it harder to read, so I think we should drop that too) and quotes the attributes.

https://www.w3.org/TR/selectors-3/#attribute-selectors

Attribute selectors are supposed to use identifiers or strings. We're using strings in these cases, so they actually should be quoted. (Stylelint really wants this to be " and not ', for consistency. So we'll go for that.)

Therefore, I'll split out the quoted attribute selectors from the "default" set and drop the others.

File: .stylelintrc.json [hunk 1]
```diff
@@ -1,97 +1,37 @@
 {
   "extends": "stylelint-config-standard-scss",
   "rules": {
-	"alpha-value-notation": null,
+	"declaration-colon-newline-after": null,
+	"selector-list-comma-newline-after": null,
```
Comment (garrett) [inline .stylelintrc.json:5]: Rules that reformat the code a lot and mess things up.

`declaration-colon-newline-after` just messes up some CSS. :frowning_face: 

`selector-list-comma-newline-after`:

```
one, two, three
```
becomes
```
one,
 two,
 three
```
(And depends on re-indention)

File: .stylelintrc.json [hunk 2]
```diff
@@ -1,97 +1,37 @@
...
-	"at-rule-semicolon-newline-after": null,
-	"color-function-notation": null,
-	"color-hex-case": null,
-	"color-hex-length": null,
+	"declaration-colon-space-before": null,
+	"declaration-empty-line-before": null,
 	"comment-empty-line-before": null,
-	"comment-whitespace-inside": null,
+	"scss/double-slash-comment-empty-line-before": null,
+	"scss/dollar-variable-colon-space-after": null,
```
Comment (garrett) [inline .stylelintrc.json:12]: These are a matter of preference.

File: .stylelintrc.json [hunk 3]
```diff
@@ -1,97 +1,37 @@
...
-	"selector-type-case": null,
-	"shorthand-property-no-redundant-values": null,
-	"keyframes-name-pattern": null,
-	"value-keyword-case": null,
-	"value-list-comma-newline-after": null,
-	"value-list-comma-space-after": null,
-	"value-list-comma-space-before": null,
-	"value-list-max-empty-lines": null,
-	"value-no-vendor-prefix": null
+	"selector-id-pattern": null
```
Comment (garrett) [inline .stylelintrc.json:35]: The rest of the rules are things that need to have complete manual intervention and need to be evaluated on a case by case basis. For now, we're turning them off.

**Comment by garrett** [comment]
Added some notes @ https://github.com/cockpit-project/cockpit/pull/17408#pullrequestreview-1021489462 on the styelintrc; they're not actionable... I would add comments in the file, but JSON does not support comments. :roll_eyes:

**Comment by garrett** [comment]
It's best to review each commit separately.

I could merge related rules together, but keeping them separate might make git history (with git blame) a bit easier.

**Pull request closed without merging**